### PR TITLE
[JENKINS-43581] - Create new custom tool button should be on top

### PR DESCRIPTION
**Pull request by Ykus**
See [JENKINS-43581](https://issues.jenkins-ci.org/browse/JENKINS-43581).

### Proposed changelog entries

* JENKINS-43581 - RFE - Show the "Add" button on the top for Tool Installations
* JENKINS-43581 - RFE - Internal:  Add the `enableTopButton` attribute to `lib/form/repeatable` Jelly control. If enabled, the "Add" button will be also shown on the top

 Adds button, for adding new entries, on top of a repeatable form (issue 43581)

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@oleg-nenashev

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
@Ykus Please do not use `@reviewbybees` if you not a CloudBees employee. Use @jenkinsci/code-reviewers  instead

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
The change impacts all `repeatable.jelly` usages, and IMHO it may cause some UX issues in cases when the list is short.  

I would advice to add a new `attrs` parameter, which enables the button on the top. And then apply it to the long layouts like tools.

The code itself looks good to me, but I am not a Javascript developer. Extra feedback would be useful

**Comment by Ykus** [comment]
@oleg-nenashev I have changed repeatable forms behavior regarding your suggestion.

File: test/src/test/resources/lib/form/RepeatableTest/testNestedRadioTopButton.jelly
Comment (oleg-nenashev) [inline test/src/test/resources/lib/form/RepeatableTest/testNestedRadioTopButton.jelly]: I am not sure about this and other headers. Was it full copy-paste of the file? If no, makes sense to update dates and to add your names

File: war/src/main/webapp/css/style.css
Comment (oleg-nenashev) [inline war/src/main/webapp/css/style.css]: Hm, is it related? I see no usages

**Review by oleg-nenashev — APPROVED** [review]
I have reviewed it yesterday. The most of the change looks good, and thanks for adding tests for it. 👍 

I have only minor comments regarding the license headers and the CSS file

File: war/src/main/webapp/css/style.css
Comment (Ykus) [inline war/src/main/webapp/css/style.css]: Good catch, Thanks.

File: test/src/test/resources/lib/form/RepeatableTest/testNestedTopButtonInner.jelly
Comment (oleg-nenashev) [inline test/src/test/resources/lib/form/RepeatableTest/testNestedTopButtonInner.jelly]: 2017- then

File: core/src/main/resources/lib/form/repeatable.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/form/repeatable.jelly]: Too vague.

**Comment by daniel-beck** [comment]
Does this just move the button, or add a second button?

What sorts of lists in Jenkins does this affect? Could this also be applied to lists of build steps in freestyle projects? IIRC there is a request to add new steps on top.

**Comment by Ykus** [comment]
It adds new (second) button above the repeatable form.

It affects only repeatable (repeatable.js) lists. Freestyle projects are made by hetero-list. I think I can make similar changes for hetero-list as I made for repeatable.

**Comment by oleg-nenashev** [comment]
I think @daniel-beck should just create a follow-up ticket for the Hetero List. I have modified the changelog entries in order to reflect which control and page are updated.

**Comment by daniel-beck** [comment]
@oleg-nenashev Alike things should behave alike. It's insane that Jenkins records who created an agent, but not any other object in Jenkins.

I'd like to take another look at this before this goes in to understand it better.

**Comment by oleg-nenashev** [comment]
ack, on hold then

**Comment by oleg-nenashev** [comment]
@daniel-beck gentle reminder. It would be nice to land it into the next LTS baseline, because it is a real problem for big tool configs

**Comment by daniel-beck** [comment]
I suppose it fixes the issue as reported by introducing yet more buttons. I feel sorry for the person who wants their stuff to appear in the middle of the list.

Besides that obvious limitation, I'm really not sure about this. As I wrote above,

> Alike things should behave alike.

And all other lists don't get this. I'm pretty sure the corresponding request for build steps is even older than this. In the case of build steps, we might be saying that it's not a good idea to have so many different ones to begin with. But personally, I'd say the same for lists of tool installations.

Not strongly opposed, but I don't think this is a good solution either.

**Comment by oleg-nenashev** [comment]
> And all other lists don't get this. I'm pretty sure the corresponding request for build steps is even older than this. In the case of build steps, we might be saying that it's not a good idea to have so many different ones to begin with

I agree though I rather see it as a follow-up ticket. Testing UX on a single control LGTM. More feedback from @jenkinsci/code-reviewers would be definitely useful

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers ping, the vacation period is almost over ;)

**Comment by oleg-nenashev** [comment]
Unassigned @daniel-beck according to his feedback in IRC, we need more votes

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers ping. Custom Tools plugin is generally broken now due to the dependency on insecure Extended Choice version, but maybe somebody is up to reviewing it

**Comment by recena** [comment]
IMHO, screenshots before/after are very helpful for this PR.

**Comment by daniel-beck** [comment]
> I agree though I rather see it as a follow-up ticket. Testing UX on a single control LGTM.

Then we need specific commitment of someone to follow up on this no later than N months from now and either revert, or implement everywhere.

**Comment by daniel-beck** [comment]
@oleg-nenashev 

I do not think the minor usability improvement here outweighs the resulting inconsistency between similar UI controls, making this a (slightly) net negative change. Hence my previous comment.

The road forward is clear IMO: Not do this, or apply it consistently.

Again, I'm not specifically _blocking_ this if you feel strongly, but it doesn't look to me that many others are interested in seeing this land based on limited feedback here.

**Comment by oleg-nenashev** [comment]
I am merging it on my own responsibility since nobody else blocks the pull request. Although it is a "minor improvement" for API users, it is a significant UX improvement for Jenkins admins who have to define many tool installations. The change is well covered by tests. Since we have much bigger barely-justified changes like XML 1.1 merged, I consider it is as a judgment call. And in this case I decide to merge this contribution from a community member instead of blocking it indefinitely.

I have created [JENKINS-49900](https://issues.jenkins-ci.org/browse/JENKINS-49900) as a follow-up to @daniel-beck's requirement. It's a newbie-friendly ticket, I will be offering it to GSoC students who are interested in UI-related project ideas.

**Comment by oleg-nenashev** [comment]
Thanks for your contribution @Ykus! My apologies for it taking so long

**Comment by recena** [comment]
@jglick 

> We are not generally investing time in “classic UI”; contributions should go toward Blue Ocean.

Long life to the classic UI.

**Pull request merged**
